packages/plugins/hmac-upstream-signature/src/index.ts

```diff
@@ -0,0 +1,136 @@
+import type { FetchAPI, GraphQLParams, Plugin as YogaPlugin } from 'graphql-yoga';
+import jsonStableStringify from 'json-stable-stringify';
+import type { OnSubgraphExecutePayload } from '@graphql-mesh/fusion-runtime';
+import type { MeshServePlugin } from '@graphql-mesh/serve-runtime';
+import { defaultPrintFn } from '@graphql-mesh/transport-common';
+import type { ExecutionRequest } from '@graphql-tools/utils';
+
+export type HMACUpstreamSignatureOptions = {
+  secret: string;
+  shouldSign?: (
+    input: Pick<OnSubgraphExecutePayload<{}>, 'subgraph' | 'subgraphName' | 'executionRequest'>,
+  ) => boolean;
+  extensionName?: string;
+  serializeExecutionRequest?: (executionRequest: ExecutionRequest) => string;
+};
+
+const DEFAULT_EXTENSION_NAME = 'hmac-signature';
+const DEFAULT_SHOULD_SIGN_FN: NonNullable<HMACUpstreamSignatureOptions['shouldSign']> = () => true;
+
+export const defaultExecutionRequestSerializer = (executionRequest: ExecutionRequest) =>
+  jsonStableStringify({
+    query: defaultPrintFn(executionRequest.document),
+    variables: executionRequest.variables,
+  });
+export const defaultParamsSerializer = (params: GraphQLParams) =>
+  jsonStableStringify({
+    query: params.query,
+    variables: params.variables,
+  });
+
+function createCryptoKey({
+  textEncoder,
+  crypto,
+  secret,
+  usages,
+}: {
+  textEncoder: TextEncoder;
+  crypto: Crypto;
+  secret: string;
+  usages: KeyUsage[];
+}): Promise<CryptoKey> {
+  return crypto.subtle.importKey(
+    'raw',
+    textEncoder.encode(secret),
+    { name: 'HMAC', hash: 'SHA-256' },
+    false,
+    usages,
+  );
+}
+
+export function useHMACUpstreamSignature(options: HMACUpstreamSignatureOptions): MeshServePlugin {
+  const shouldSign = options.shouldSign || DEFAULT_SHOULD_SIGN_FN;
+  const extensionName = options.extensionName || DEFAULT_EXTENSION_NAME;
+  const serializeExecutionRequest =
+    options.serializeExecutionRequest || defaultExecutionRequestSerializer;
+  let key$: Promise<CryptoKey>;
+  let fetchAPI: FetchAPI;
+  let textEncoder: TextEncoder;
+
+  return {
+    onYogaInit({ yoga }) {
+      fetchAPI = yoga.fetchAPI;
+    },
+    onSubgraphExecute({ subgraphName, subgraph, executionRequest, setExecutionRequest }) {
+      if (shouldSign({ subgraphName, subgraph, executionRequest })) {
+        textEncoder ||= new fetchAPI.TextEncoder();
+        key$ ||= createCryptoKey({
+          textEncoder,
+          crypto: fetchAPI.crypto,
+          secret: options.secret,
+          usages: ['sign'],
+        });
```

Comment:
We don't want to load the key eagerly ? We want to wait for the first request to be handled before knowing there is an error in the key format or that it's missing ?

Reply:
It's not really fetching in this case, the key is loaded from an env var. 
It's just the crypto API that is async.

Reply:
Yes, but this means that if the key is malformed or missing, it will be known only once the server is started. The server will appear as running while it is in fact enable to serve any requests

Reply:
good point. fixed.